### PR TITLE
feat(language): Add centralized ExprEvaluator for closure expression resolution

### LINKED ISSUE #205: [Bug] parse shape failed when shape is not list of const
### Component

Parser

### Description

```
def test_elementwise_add(dtype = pl.FP32, shape = [128, 128], tile_shape = [64, 64]):
    @pl.function
    def kernel_add(
        t: pl.Tensor[shape, dtype], out: pl.Tensor[shape, dtype]
    ) -> pl.Tensor[shape, dtype]:
        a: pl.Tile[tile_shape, dtype] = pl.block.load(t, offsets=[0, 0], shapes=tile_shape)
        b: pl.Tile[tile_shape, dtype] = pl.add(a, 5)
        result: pl.Tensor[shape, dtype] = pl.block.store(
            b, offsets=[0, 0], shapes=tile_shape, output_tensor=out
        )
        return result

    t = torch.ones(shape, dtype=torch.float32)
    out = torch.zeros(shape, dtype=torch.float32)
    kernel_add(t, out)
```

### Steps to Reproduce

```markdown
execute the test snippets
```

### Expected Behavior

function parse correctly

### Actual Behavior

parse failed

### Git Commit ID

LATEST

### NPU Kind

N/A (not hardware-specific)

### Host Architecture

x86_64

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary
- Add `ExprEvaluator` class that uses Python `eval()` with restricted builtins whitelist to resolve arbitrary expressions (arithmetic, subscripts, builtin calls) against closure variables
- Refactor `TypeResolver` to accept `ExprEvaluator` instead of raw `closure_vars` dict, enabling expression-based shape dimensions like `[base * 2, config["cols"]]`
- Refactor `ASTParser._parse_op_kwargs` to use `ExprEvaluator` for resolving closure variables in keyword arguments (e.g., `shapes=tile_shape`)

## Testing
- [x] 28 new `ExprEvaluator` unit tests (arithmetic, builtins, subscripts, error cases)
- [x] 33 new user-perspective integration tests (expression shapes, closure kwargs, edge cases, factory patterns)
- [x] All 308 parser tests pass
- [x] Full suite of 1456 tests pass with no regressions
- [x] Pre-commit hooks pass (ruff check, ruff format, pyright)

## Related Issues
Fixes #205

### CODE REVIEW

File: python/pypto/language/parser/expr_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+        Raises:
+            ParserTypeError: If expression cannot be evaluated
+        """
+        span = self._get_span(node)
+        expr_str = ast.unparse(node)
+
+        try:
+            code = compile(ast.Expression(body=node), "<pypto-eval>", "eval")
+            return eval(code, {"__builtins__": _SAFE_BUILTINS}, dict(self.closure_vars))  # noqa: S307
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/expr_evaluator.py:88]: Acknowledged. Added an explicit security comment to the eval() call explaining the threat model: the DSL parser is not a sandbox — users already have full control of the Python process, so closure_vars come from the user's own scope. The builtins whitelist prevents accidental access to dangerous builtins. A full AST interpreter is a good hardening step for the future but out of scope for this PR.

File: python/pypto/language/parser/ast_parser.py [hunk 2]
```diff
@@ -1325,25 +1330,51 @@ def _parse_op_kwargs(self, call: ast.Call) -> dict[str, Any]:
...
+    def _resolve_name_kwarg(self, value: ast.Name) -> Any:
+        """Resolve a Name kwarg value via scope lookup or closure eval."""
+        if value.id in ["True", "False"]:
+            return value.id == "True"
+        if self.scope_manager.lookup_var(value.id) is not None:
+            return self.parse_expression(value)  # IR var from scope
+        success, result = self.expr_evaluator.try_eval_expr(value)
+        if success:
+            return result
+        return self.parse_expression(value)  # raises UndefinedVariableError
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/ast_parser.py]: Fixed in 954e68a — `_resolve_name_kwarg` now calls `self.expr_evaluator.eval_expr(value)` directly instead of falling back to `parse_expression`.

File: python/pypto/language/parser/ast_parser.py [hunk 3]
```diff
@@ -1325,25 +1330,51 @@ def _parse_op_kwargs(self, call: ast.Call) -> dict[str, Any]:
...
+
+    def _resolve_attribute_kwarg(self, value: ast.Attribute) -> Any:
+        """Resolve an Attribute kwarg value (e.g., pl.FP32, config.field)."""
+        try:
+            return self.type_resolver.resolve_dtype(value)
+        except ParserTypeError:
+            success, result = self.expr_evaluator.try_eval_expr(value)
+            if success:
+                return result
+            raise
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/ast_parser.py]: Fixed in 9f9710e — `_resolve_attribute_kwarg` now calls `self.expr_evaluator.eval_expr(value)` directly in the except block, so failures surface expression-specific errors instead of the misleading dtype error.

File: tests/ut/language/parser/test_type_resolver.py [hunk 4]
```diff
@@ -9,22 +9,33 @@
...
 import pypto.language as pl
 import pytest
 from pypto.language.parser.diagnostics import ParserTypeError
+from pypto.language.parser.expr_evaluator import ExprEvaluator
 from pypto.language.parser.type_resolver import TypeResolver
 from pypto.language.typing.dynamic import DynVar
 from pypto.pypto_core import DataType, ir
 
 
+def _make_resolver(closure_vars: dict | None = None, scope_lookup=None) -> TypeResolver:
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/language/parser/test_type_resolver.py]: Fixed in 9f9710e — added type hints to `_make_resolver` with `Callable` and `Any` imports.

File: python/pypto/language/parser/expr_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        except NameError as e:
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/expr_evaluator.py:89]: Acknowledged — added an explicit security comment documenting the threat model. The DSL parser is not a sandbox; users already have full Python process control. The builtins whitelist is defense-in-depth against accidental access, not a security boundary. A full AST interpreter would be a good future hardening step.

File: python/pypto/language/parser/ast_parser.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/ast_parser.py]: Fixed in 9f9710e — `_resolve_attribute_kwarg` now uses `eval_expr` directly to surface expression-specific errors.

File: python/pypto/language/parser/type_resolver.py [hunk 6]
```diff
@@ -367,21 +454,10 @@ def _resolve_shape_dim(self, name_node: ast.Name) -> int | ir.Expr:
...
-                return ir.Var(
-                    value.name,
-                    ir.ScalarType(DataType.INT64),
-                    span,
-                )
-            raise ParserTypeError(
-                f"Shape variable '{name}' must be int or pl.dynamic(), got {type(value).__name__}",
-                span=span,
-            )
+        # 1. Try evaluating via ExprEvaluator (handles closure vars)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/language/parser/type_resolver.py]: Fixed in 9f9710e — added a fast-path direct dict lookup in `_resolve_shape_dim` that checks `closure_vars` directly for simple names, avoiding compile+eval overhead for the common case.
